e2e/libs/page-manager.ts

```diff
@@ -0,0 +1,251 @@
+import {
+  canLogResponse,
+  describeJsHandle,
+  getRequestData,
+  isApiFailure,
+  isWorkbenchRequest,
+  logError,
+  shouldSkipApiResponseBody,
+  transformResponseBody
+} from './page-events-helper';
+import { logger } from './logger';
+import { defaultLaunchOptions } from './page-options';
+import fs from 'fs-extra';
+import { signIn, signInWithAccessToken } from 'utils/test-utils';
+import { savePageToFile, takeScreenshot } from 'utils/save-file-utils';
+
+const failScreenshotDir = 'logs/screenshot';
+const failHtmlDir = 'logs/html';
+const { CIRCLE_BUILD_NUM } = process.env;
+const userAgent =
+  'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) ' +
+  'Chrome/80.0.3987.149 Safari/537.36' +
+  (CIRCLE_BUILD_NUM ? ` (circle-build-number/${CIRCLE_BUILD_NUM})` : '');
+
+/**
+ * Launch new Chrome.
+ * @param launchOpts: {@link LaunchOptions} New browser launch options.
+ * @return {@link Browser}
+ */
+export const launchBrowser = async (launchOpts?: LaunchOptions): Promise<Browser> => {
+  return launch(launchOpts ? launchOpts : defaultLaunchOptions);
+};
+
+/**
+ * Launch new incognito page.
+ * @param browser: {@link Browser}
+ * @return Page
+ */
+export const launchPage = async (browser: Browser): Promise<Page> => {
+  const incognitoPage = await browser.createIncognitoBrowserContext().then((context) => context.newPage());
+  await initPageBeforeTest(incognitoPage);
+  return incognitoPage;
+};
+
+export const withBrowser =
+  (launchOpts?: LaunchOptions) =>
+  async (testFn: (browser: Browser) => Promise<void>): Promise<void> => {
+    const browser = await launchBrowser(launchOpts);
+    try {
+      await testFn(browser);
+    } finally {
+      await browser
+        .close()
+        .then(() => logger.info('Closing browser'))
+        .catch((err: Error) => {
+          console.error(`Unable to close browser. Error message: ${err.message}`);
+        });
+    }
+  };
+
+/**
+ * Launch new browser and incognito page. Opens Workbench Login page.
+ * @param launchOpts: {@link LaunchOptions} New browser launch options.
+ */
+export const withPage =
+  (launchOpts?: LaunchOptions) =>
+  async (testFn: (page: Page, browser: Browser) => Promise<void>): Promise<void> => {
+    await withBrowser(launchOpts)(async (browser) => {
+      const incognitoPage = await browser.createIncognitoBrowserContext().then((context) => context.newPage());
+      try {
+        await initPageBeforeTest(incognitoPage);
+        await testFn(incognitoPage, browser);
+      } catch (err) {
+        if (err instanceof Error) {
+          logger.error(err.message);
+        }
+        // Take screenshot and save html contents immediately after failure.
+        await fs.ensureDir(failScreenshotDir);
+        await fs.ensureDir(failHtmlDir);
+        await takeScreenshot(incognitoPage, `${__SPEC_NAME__}.png`);
+        await savePageToFile(incognitoPage, `${__SPEC_NAME__}.html`);
```

Comment:
Take screenshot and save html contents on failure. Couldn't do that in `puppteer-custom-environment.ts` anymore because the page object is closed here.